### PR TITLE
Add python code for Subtree Queries in Euler Tour module

### PR DESCRIPTION
**Pull request by brebenelmihnea**
_If any of the below don't apply to this Pull Request, mark the checkbox as done._

- [x] I have tested my code.
- [x] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [x] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions).
  - I understand that if it is clear that I have not attempted to follow these conventions, my PR will be closed.
  - If changes are requested, I will re-request a review after addressing them.

### CODE REVIEW

**Comment by SansPapyrus683** [comment]
:skull: thought i asked you to make a solution for a problem that didn't already have one
this one is completely, 100% fine tho

File: content/4_Gold/Tree_Euler.mdx [hunk 1]
```diff
@@ -398,7 +419,81 @@ traversal is now complete.
...
+		dfs(son, node)
+	en[node] = pos - 1
+
+
+for i in range(n - 1):
+	x, y = map(int, input().split())
+	g[x].append(y)
+	g[y].append(x)
+
+sys.setrecursionlimit(200002)
```
Comment (SansPapyrus683) [inline content/4_Gold/Tree_Euler.mdx:481]:
```suggestion
sys.setrecursionlimit(10 ** 9)
```

**Comment by brebenelmihnea** [comment]
> skull thought i asked you to make a solution for a problem that didn't already have one this one is completely, 100% fine tho

I looked in the entire guide. It's really hard to find a problem without any solution. This one didn't have any python code .

File: content/4_Gold/Tree_Euler.mdx [hunk 2]
```diff
@@ -39,6 +39,27 @@ size of $i$.
 
 <LanguageSection>
 
+<PySection>
```
[the remark this answers is not in the document]
Comment (brebenelmihnea) [reply content/4_Gold/Tree_Euler.mdx]: Done

File: content/4_Gold/Tree_Euler.mdx [hunk 3]
```diff
@@ -39,6 +39,27 @@ size of $i$.
...
+
+
+def dfs(node: int, parent: int):
+	global pos
+	st[node] = pos
+	pos += 1
+	for son in g[node]:
+		if son == parent:
+			continue
+		dfs(son, node)
```
[the remark this answers is not in the document]
Comment (brebenelmihnea) [reply content/4_Gold/Tree_Euler.mdx]: Done

File: content/4_Gold/Tree_Euler.mdx [hunk 4]
```diff
@@ -398,7 +419,81 @@ traversal is now complete.
...
+<PySection>
+
+```py
+import sys
+
+n, m = map(int, input().split())
+
+v = list(map(int, input().split()))
+# BeginCodeSnip{Segment Tree}
+class Segtree:
```
[the remark this answers is not in the document]
Comment (brebenelmihnea) [reply content/4_Gold/Tree_Euler.mdx]: Done

File: content/4_Gold/Tree_Euler.mdx (hunk 1 above)
Comment (brebenelmihnea) [reply content/4_Gold/Tree_Euler.mdx]: Done

File: content/4_Gold/Tree_Euler.mdx [hunk 5]
```diff
@@ -39,6 +39,27 @@ size of $i$.
 
 <LanguageSection>
 
+<PySection>
+
+```py
+g = [[] for i in range(200005)]
+st = [0 for i in range(200005)]
+en = [0 for i in range(200005)]
```
[the remark this answers is not in the document]
Comment (brebenelmihnea) [reply content/4_Gold/Tree_Euler.mdx]: Done

File: content/4_Gold/Tree_Euler.mdx (hunk 4 above)
[the remark this answers is not in the document]
Comment (brebenelmihnea) [reply content/4_Gold/Tree_Euler.mdx]: Done

File: content/4_Gold/Tree_Euler.mdx [hunk 6]
```diff
@@ -398,7 +419,81 @@ traversal is now complete.
...
+		return self.query(2 * node, left, mid, x, y) + self.query(
+			2 * node + 1, mid + 1, right, x, y
+		)
+
+
+# EndCodeSnip
+
+g = [[] for i in range(200005)]
+st = [0 for i in range(200005)]
+en = [0 for i in range(200005)]
```
[the remark this answers is not in the document]
Comment (brebenelmihnea) [reply content/4_Gold/Tree_Euler.mdx]: Done

File: content/4_Gold/Tree_Euler.mdx (hunk 5 above)
[the remark this answers is not in the document]
Comment (SansPapyrus683) [reply content/4_Gold/Tree_Euler.mdx]: not fixed

File: content/4_Gold/Tree_Euler.mdx [hunk 7]
```diff
@@ -71,6 +71,28 @@ public static void dfs(int i, int p) {
 
 </JavaSection>
 
+<PySection>
+
+```py
+g = [[] for i in range(200005)]
+st = [0 for i in range(200005)]
+en = [0 for i in range(200005)]
+pos = 1
```
Comment (envyaims) [inline content/4_Gold/Tree_Euler.mdx:80]: use `timer` like in the other languages of the module

File: content/4_Gold/Tree_Euler.mdx [hunk 8]
```diff
@@ -542,6 +564,86 @@ public class Main {
...
+			dfs(son, node)
+	en[node] = pos - 1
+
+
+for i in range(n - 1):
+	x, y = map(int, input().split())
+	g[x].append(y)
+	g[y].append(x)
+
+sys.setrecursionlimit(10**9)
```
Comment (envyaims) [inline content/4_Gold/Tree_Euler.mdx:634]: why is this here? move this to the top?

File: content/4_Gold/Tree_Euler.mdx [hunk 9]
```diff
@@ -542,6 +564,86 @@ public class Main {
...
+dfs(1, 1)
+sg = SegmentTree(v, n)
+
+for i in range(m):
+	vals = list(map(int, input().split()))
```
Comment (envyaims) [inline content/4_Gold/Tree_Euler.mdx:639]: `query` would be a better name

File: content/4_Gold/Tree_Euler.mdx [hunk 10]
```diff
@@ -542,6 +564,86 @@ public class Main {
 
 </JavaSection>
 
+<PySection>
+
+```py
+import sys
+
+NMAX = 200005
+n, m = map(int, input().split())
```
Comment (envyaims) [inline content/4_Gold/Tree_Euler.mdx:573]: why read as `m`? Isn't it 'q' in the official statement?

**Review by envyaims — CHANGES_REQUESTED** [review]

File: content/4_Gold/Tree_Euler.mdx (hunk 7 above)
Comment (brebenelmihnea) [reply content/4_Gold/Tree_Euler.mdx]: I changed it to timer

File: content/4_Gold/Tree_Euler.mdx (hunk 8 above)
Comment (brebenelmihnea) [reply content/4_Gold/Tree_Euler.mdx]: I moved it to the top

File: content/4_Gold/Tree_Euler.mdx (hunk 9 above)
Comment (brebenelmihnea) [reply content/4_Gold/Tree_Euler.mdx]: Done

File: content/4_Gold/Tree_Euler.mdx (hunk 10 above)
Comment (brebenelmihnea) [reply content/4_Gold/Tree_Euler.mdx]: I changed $m$ to $q$

File: content/4_Gold/Tree_Euler.mdx (hunk 5 above)
[the remark this answers is not in the document]
Comment (brebenelmihnea) [reply content/4_Gold/Tree_Euler.mdx]: Then what do you mean by magic numbers ?

[the remark this answers is not in the document]
Comment (SansPapyrus683) [reply content/4_Gold/Tree_Euler.mdx]: https://stackoverflow.com/questions/47882

**Review by envyaims — APPROVED** [review]

File: content/4_Gold/Tree_Euler.mdx (hunk 5 above)
[the remark this answers is not in the document]
Comment (brebenelmihnea) [reply content/4_Gold/Tree_Euler.mdx]: image
Yes, I did it the first time.

File: content/4_Gold/Tree_Euler.mdx [hunk 11]
```diff
@@ -542,6 +563,86 @@ public class Main {
...
+
+sys.setrecursionlimit(10**9)
+
+NMAX = 200005
+n, q = map(int, input().split())
+
+v = list(map(int, input().split()))
+
+
+class SegmentTree:
```
Comment (SansPapyrus683) [inline content/4_Gold/Tree_Euler.mdx:579]: put in codesnip

File: content/4_Gold/Tree_Euler.mdx [hunk 12]
```diff
@@ -542,6 +563,86 @@ public class Main {
...
+	def __init__(self, arr: list[int], n: int) -> None:
```
Comment (SansPapyrus683) [inline content/4_Gold/Tree_Euler.mdx:580]: use `List` from `typing`

File: content/4_Gold/Tree_Euler.mdx [hunk 13]
```diff
@@ -542,6 +563,86 @@ public class Main {
...
+		self.arr = arr
+		self.n = n
+		self.tree = [0] * (2 * n)
+		for ii in range(n):
+			self.tree[n + ii] = arr[ii]
+		for ii in reversed(range(n)):
+			self.tree[ii] = self.tree[ii << 1] + self.tree[(ii << 1) | 1]
+
+	def add(self, i: int, v: int) -> None:
```
Comment (SansPapyrus683) [inline content/4_Gold/Tree_Euler.mdx:589]:
```suggestion
	def add(self, i: int, v: int):
```

File: content/4_Gold/Tree_Euler.mdx [hunk 14]
```diff
@@ -542,6 +563,86 @@ public class Main {
...
 </JavaSection>
 
+<PySection>
+
+```py
+import sys
+
+sys.setrecursionlimit(10**9)
+
+NMAX = 200005
```
Comment (SansPapyrus683) [inline content/4_Gold/Tree_Euler.mdx:573]: is the +5 necessary

Comment (SansPapyrus683) [reply content/4_Gold/Tree_Euler.mdx:573]: also rename to MAX_N

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: content/4_Gold/Tree_Euler.mdx (hunk 11 above)
Comment (brebenelmihnea) [reply content/4_Gold/Tree_Euler.mdx:579]: Done

File: content/4_Gold/Tree_Euler.mdx (hunk 13 above)
Comment (brebenelmihnea) [reply content/4_Gold/Tree_Euler.mdx:589]: Done

File: content/4_Gold/Tree_Euler.mdx (hunk 14 above)
Comment (brebenelmihnea) [reply content/4_Gold/Tree_Euler.mdx]: I changed to MAX_N and removed +5

File: content/4_Gold/Tree_Euler.mdx (hunk 12 above)
Comment (brebenelmihnea) [reply content/4_Gold/Tree_Euler.mdx]: What do you mean ? Isn't the type hint right ?

Comment (brebenelmihnea) [reply content/4_Gold/Tree_Euler.mdx]: I used list

Comment (SansPapyrus683) [reply content/4_Gold/Tree_Euler.mdx]: generic type hints aren't supported till 3.9 i think, and usaco uses 3.6
so just `from typing import List` and use that instead

Comment (brebenelmihnea) [reply content/4_Gold/Tree_Euler.mdx]: image
It's supported since 3.5

Comment (brebenelmihnea) [reply content/4_Gold/Tree_Euler.mdx]: Problem solved :))

File: content/4_Gold/Tree_Euler.mdx [hunk 15]
```diff
@@ -542,6 +563,88 @@ public class Main {
...
+			if r & 1:
+				r -= 1
+				sum += self.tree[r]
+			l >>= 1
+			r >>= 1
+		return sum
+
+
+# EndCodeSnip
+
```
Comment (SansPapyrus683) [inline content/4_Gold/Tree_Euler.mdx:614]:
```suggestion
```

File: content/4_Gold/Tree_Euler.mdx [hunk 16]
```diff
@@ -542,6 +563,88 @@ public class Main {
...
 </JavaSection>
 
+<PySection>
+
+```py
+import sys
+
+sys.setrecursionlimit(10**9)
+
+MAX_N = 200000
```
Comment (SansPapyrus683) [inline content/4_Gold/Tree_Euler.mdx:573]: did you test this code

File: content/4_Gold/Tree_Euler.mdx (hunk 13 above)
Comment (SansPapyrus683) [reply content/4_Gold/Tree_Euler.mdx:589]: i don't see it changed

File: content/4_Gold/Tree_Euler.mdx [hunk 17]
```diff
@@ -542,6 +563,88 @@ public class Main {
...
+n, q = map(int, input().split())
+
+v = list(map(int, input().split()))
+
+# BeginCodeSnip{Segment Tree}
```
Comment (SansPapyrus683) [inline content/4_Gold/Tree_Euler.mdx:578]:
```suggestion
# BeginCodeSnip{Segment Tree (from PURS module)}
```

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: content/4_Gold/Tree_Euler.mdx (hunk 13 above)
Comment (brebenelmihnea) [reply content/4_Gold/Tree_Euler.mdx]: I removed it

File: content/4_Gold/Tree_Euler.mdx (hunk 17 above)
Comment (brebenelmihnea) [reply content/4_Gold/Tree_Euler.mdx]: I added it

File: content/4_Gold/Tree_Euler.mdx (hunk 15 above)
Comment (brebenelmihnea) [reply content/4_Gold/Tree_Euler.mdx:612]: I removed the empty lines

File: content/4_Gold/Tree_Euler.mdx (hunk 16 above)
Comment (brebenelmihnea) [reply content/4_Gold/Tree_Euler.mdx]: I added +1

File: content/4_Gold/Tree_Euler.mdx [hunk 18]
```diff
@@ -71,6 +71,27 @@ public static void dfs(int i, int p) {
 
 </JavaSection>
 
+<PySection>
+
+```py
+g = [[] for i in range(NMAX)]
+st = [0 for i in range(NMAX)]
+en = [0 for i in range(NMAX)]
```
Comment (SansPapyrus683) [inline content/4_Gold/Tree_Euler.mdx:79]: MAX_N here as well

File: content/4_Gold/Tree_Euler.mdx [hunk 19]
```diff
@@ -542,6 +563,88 @@ public class Main {
...
+	global timer
+	st[node] = timer
+	timer += 1
+	for son in g[node]:
+		if son != parent:
+			dfs(son, node)
+	en[node] = timer - 1
+
+
+for i in range(n - 1):
```
Comment (SansPapyrus683) [inline content/4_Gold/Tree_Euler.mdx:631]:
```suggestion
for _ in range(n - 1):
```

File: content/4_Gold/Tree_Euler.mdx [hunk 20]
```diff
@@ -542,6 +563,88 @@ public class Main {
...
 </JavaSection>
 
+<PySection>
+
+```py
+import sys
+
+sys.setrecursionlimit(10**9)
+
+MAX_N = 200001
```
Comment (SansPapyrus683) [inline content/4_Gold/Tree_Euler.mdx:573]: 0-index the nodes when reading input & making this 2 * 10 ** 5

File: content/4_Gold/Tree_Euler.mdx [hunk 21]
```diff
@@ -542,6 +563,88 @@ public class Main {
...
+			dfs(son, node)
+	en[node] = timer - 1
+
+
+for i in range(n - 1):
+	x, y = map(int, input().split())
+	g[x].append(y)
+	g[y].append(x)
+
+dfs(1, 1)
```
Comment (SansPapyrus683) [inline content/4_Gold/Tree_Euler.mdx:636]: it would also necessitate changing this to dfs(0, 0)

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: content/4_Gold/Tree_Euler.mdx (hunk 18 above)
Comment (brebenelmihnea) [reply content/4_Gold/Tree_Euler.mdx]: Done.

File: content/4_Gold/Tree_Euler.mdx (hunk 19 above)
Comment (brebenelmihnea) [reply content/4_Gold/Tree_Euler.mdx]: Done.

File: content/4_Gold/Tree_Euler.mdx (hunk 20 above)
Comment (brebenelmihnea) [reply content/4_Gold/Tree_Euler.mdx]: Done.

File: content/4_Gold/Tree_Euler.mdx (hunk 21 above)
Comment (brebenelmihnea) [reply content/4_Gold/Tree_Euler.mdx]: Done.

File: content/4_Gold/Tree_Euler.mdx [hunk 22]
```diff
@@ -542,6 +563,89 @@ public class Main {
 
 </JavaSection>
 
+<PySection>
+
+```py
+import sys
+
+sys.setrecursionlimit(10**9)
```
Comment (SansPapyrus683) [inline content/4_Gold/Tree_Euler.mdx:571]:
```suggestion
sys.setrecursionlimit(10 ** 9)
```

File: content/4_Gold/Tree_Euler.mdx [hunk 23]
```diff
@@ -542,6 +563,89 @@ public class Main {
...
+	timer += 1
+	for son in g[node]:
+		if son != parent:
+			dfs(son, node)
+	en[node] = timer - 1
+
+
+for _ in range(n - 1):
+	x, y = map(int, input().split())
+	x -= 1, y -= 1
```
Comment (SansPapyrus683) [inline content/4_Gold/Tree_Euler.mdx:633]:
```suggestion
	x -= 1
	y -= 1
```
my guy did you test this

File: content/4_Gold/Tree_Euler.mdx [hunk 24]
```diff
@@ -542,6 +563,89 @@ public class Main {
...
+	g[x].append(y)
+	g[y].append(x)
+
+dfs(0, 0)
+sg = SegmentTree(v, n)
+
+for i in range(q):
```
Comment (SansPapyrus683) [inline content/4_Gold/Tree_Euler.mdx:640]:
```suggestion
for _ in range(q):

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: content/4_Gold/Tree_Euler.mdx (hunk 22 above)
Comment (brebenelmihnea) [reply content/4_Gold/Tree_Euler.mdx]: Done.

File: content/4_Gold/Tree_Euler.mdx (hunk 23 above)
Comment (brebenelmihnea) [reply content/4_Gold/Tree_Euler.mdx]: Oh, sorry. I thought it works in Python :))

File: content/4_Gold/Tree_Euler.mdx (hunk 24 above)
Comment (brebenelmihnea) [reply content/4_Gold/Tree_Euler.mdx]: Done

File: content/4_Gold/Tree_Euler.mdx (hunk 22 above)
Comment (SansPapyrus683) [reply content/4_Gold/Tree_Euler.mdx:571]: no it isn't

Comment (brebenelmihnea) [reply content/4_Gold/Tree_Euler.mdx:571]: I think the `pre-commit.ci` bot  keeps changing the file.

File: content/4_Gold/Tree_Euler.mdx [hunk 25]
```diff
@@ -542,6 +563,90 @@ public class Main {
...
+
+for _ in range(n - 1):
+	x, y = map(int, input().split())
+	x -= 1
+	y -= 1
+	g[x].append(y)
+	g[y].append(x)
+
+dfs(0, 0)
+sg = SegmentTree(v, n)
```
Comment (SansPapyrus683) [inline content/4_Gold/Tree_Euler.mdx:639]: rename to `st`

Comment (brebenelmihnea) [reply content/4_Gold/Tree_Euler.mdx]: I renamed it

**Comment by SansPapyrus683** [comment]
image
my man did you test your code????
